### PR TITLE
fix(text-direction): respect @scope boundaries

### LINKED ISSUE #1057: text-direction: @scope boundaries not respected when scanning direction rules
## Context

Flagged by review on #993 (`packages/components/src/_internal/text-direction.ts`, `matchesDirectionStyleRuleList`).

When a direction declaration is nested in `@scope`, e.g.:

```css
@scope (.theme) {
  .shell {
    direction: ltr;
  }
}
```

the recursive rule walk descends into the scope's nested rules and calls
`element.matches('.shell')` without first checking whether `element` is
inside the scoping root (`.theme`) at all. An unscoped `.shell` ancestor
outside any `.theme` still matches the inner selector here, even though the
browser does not apply the scoped rule to it.

## Impact

Under an RTL `LocaleProvider` on an otherwise LTR document, an unscoped
ancestor's default computed LTR direction can be mistaken for an intentional
override, causing MegaMenu (or any consumer of `resolveTextDirection()`) to
render and navigate in the wrong direction.

## Suggested fix

When encountering a `CSSScopeRule` (or an unknown rule whose `cssText`
starts with `@scope`), determine its scoping root selector(s) and confirm
`element` is a descendant of a matching root (respecting an optional
`to (...)` limit) before recursing into its nested rules — mirroring how
`isConditionalRuleActive()` already gates `@container`/`@media`/`@supports`.

## Scope

- `packages/components/src/_internal/text-direction.ts`
- Needs unit coverage for: element inside the scope root, element outside
  it, and a scope with a `to (...)` limit.

### PR DESCRIPTION
**Pull request by stevekinney**
## Summary
- respect CSS @scope roots and optional limits while scanning direction rules
- fail closed for malformed or unsupported scope selectors
- add regression coverage for scoped, nested, relative, and selector-list cases

## Validation
- `bun test packages/components/src/_internal/text-direction.test.ts`
- `bun run --filter=@lostgradient/cinder typecheck`
- `bun run --filter=@lostgradient/cinder lint`
- `bun run --filter=@lostgradient/cinder components:check`

Closes #1057

### CODE REVIEW

File: packages/components/src/_internal/text-direction-css.ts [hunk 1]
```diff
@@ -163,6 +177,311 @@ function isConditionalRuleActive(
...
+        const marker = 'data-cinder-scope-root';
+        const previous = root.getAttribute(marker);
+        try {
+          root.setAttribute(marker, 'true');
+          const boundSelector = selector.replaceAll(':scope', `[${marker}="true"]`);
+          if (element.matches(boundSelector)) return true;
+        } finally {
+          if (previous === null) root.removeAttribute(marker);
+          else root.setAttribute(marker, previous);
+        }
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Removed live-root attribute fallback from normal path; native :scope uses querySelectorAll directly, while unsupported engines evaluate against a detached clone, so observed DOM is never mutated.

File: packages/components/src/_internal/text-direction-css.ts [hunk 2]
```diff
@@ -163,6 +177,311 @@ function isConditionalRuleActive(
...
+
+function matchesScopedSelector(
+  element: HTMLElement,
+  selector: string,
+  scopes: readonly ActiveScope[],
+): boolean {
+  if (!selector.includes(':scope')) return matchesSelectorSafely(element, selector);
+  if (scopes.length === 0) return false;
+  for (const scope of scopes) {
+    for (const root of scope.roots) {
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: matchesScopedSelector now evaluates only the innermost active scope, with regression coverage for nested scopes.

File: packages/components/src/_internal/text-direction-css.ts [hunk 3]
```diff
@@ -163,6 +177,311 @@ function isConditionalRuleActive(
...
+    if (
+      root &&
+      prelude.limitSelectors &&
+      isWithinScopeLimit(element, root, prelude.limitSelectors)
+    ) {
+      continue;
+    }
+    if (root) activeRoots.push(root);
+    else if (!prelude.limitSelectors || !findScopeMatches(element, prelude.limitSelectors).length)
+      activeRoots.push(element.ownerDocument.documentElement);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Implicit unrooted scopes now derive their root from the owning inline style element parent, and reject targets outside that root; regression coverage added.

File: packages/components/src/_internal/text-direction-css.ts [hunk 4]
```diff
@@ -163,6 +177,311 @@ function isConditionalRuleActive(
...
+    for (const root of scope.roots) {
+      try {
+        if (Array.from(root.querySelectorAll(selector)).includes(element)) return true;
+        // Some CSSOM implementations do not implement :scope on querySelectorAll.
+        // Bind it explicitly to a temporary attribute and use the target's matcher.
+        const marker = 'data-cinder-scope-root';
+        const previous = root.getAttribute(marker);
+        try {
+          root.setAttribute(marker, 'true');
+          const boundSelector = selector.replaceAll(':scope', `[${marker}="true"]`);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Relative-selector fallback now runs on a detached clone rather than mutating the live scope root, preventing observer feedback loops.

File: packages/components/src/_internal/text-direction-css.ts [hunk 5]
```diff
@@ -163,6 +177,311 @@ function isConditionalRuleActive(
...
+  }
+}
+
+function matchesScopedSelector(
+  element: HTMLElement,
+  selector: string,
+  scopes: readonly ActiveScope[],
+): boolean {
+  if (!selector.includes(':scope')) return matchesSelectorSafely(element, selector);
+  if (scopes.length === 0) return false;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Added token-aware, case-insensitive :scope detection that ignores quoted attribute values; regression coverage covers :SCOPE and an attribute containing ':scope'.

**Comment by stevekinney** [comment]
@codex review Please perform a fresh exact-head review of d9404bc0d0a448810c3c9bcc9c2910e57842474d.

File: packages/components/src/_internal/text-direction-css.ts [hunk 6]
```diff
@@ -48,6 +57,13 @@ function isActiveStyleSheet(sheet: CSSStyleSheet): boolean {
   return matchMedia(mediaText).matches;
 }
 
+function getImplicitScopeRoot(sheet: CSSStyleSheet): Element | null {
+  const ownerNode = Reflect.get(sheet, 'ownerNode');
+  return ownerNode instanceof Element && ownerNode.tagName === 'STYLE'
+    ? ownerNode.parentElement
+    : null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 1f60d4579. Implicit @scope roots now preserve a style element's enclosing ShadowRoot when it has no parentElement, with regression coverage for a shadow-root style sheet.

File: packages/components/src/_internal/text-direction-css.ts [hunk 7]
```diff
@@ -163,6 +218,371 @@ function isConditionalRuleActive(
...
+}
+
+function isWithinScopeLimit(
+  element: HTMLElement,
+  root: Element,
+  limitSelectors: string[],
+): boolean {
+  let current: Element | null = element;
+  while (current) {
+    if (limitSelectors.some((selector) => matchesSelectorSafely(current!, selector))) return true;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 1f60d4579. Limit selectors containing :scope are evaluated against each resolved active scope root, with regression coverage for @scope (.theme) to (:scope > .stop).

File: packages/components/src/_internal/text-direction-css.ts [hunk 8]
```diff
@@ -163,6 +221,378 @@ function isConditionalRuleActive(
...
+  }
+}
+
+function matchesScopedSelector(
+  element: Element,
+  selector: string,
+  scopes: readonly ActiveScope[],
+): boolean {
+  if (!hasScopePseudoClass(selector)) return matchesSelectorSafely(element, selector);
+  if (scopes.length === 0) return false;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Adjudicated as fail-closed by design: the resolver only treats :scope as meaningful inside an active @scope context. Added regression coverage proving top-level :scope returns the caller fallback, preventing a false styling hint.

File: packages/components/src/_internal/text-direction-css.ts [hunk 9]
```diff
@@ -48,6 +57,16 @@ function isActiveStyleSheet(sheet: CSSStyleSheet): boolean {
   return matchMedia(mediaText).matches;
 }
 
+type ScopeRoot = Element | ShadowRoot;
+
+function getImplicitScopeRoot(sheet: CSSStyleSheet): ScopeRoot | null {
+  const ownerNode = Reflect.get(sheet, 'ownerNode');
+  if (!(ownerNode instanceof Element) || ownerNode.tagName !== 'STYLE') return null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in fbe88f53c. Ownerless and non-STYLE sheets now inherit the target ShadowRoot as the implicit root; regression coverage covers constructed/adopted-style-like sheets.

File: packages/components/src/_internal/text-direction-css.ts [hunk 10]
```diff
@@ -163,6 +221,378 @@ function isConditionalRuleActive(
...
+}
+
+function findScopeMatches(element: HTMLElement, selectors: string[]): Element[] {
+  if (selectors.length === 0) return [];
+  const matches: Element[] = [];
+  let current: Element | null = element;
+  while (current) {
+    const currentElement: Element = current;
+    if (selectors.some((selector) => matchesSelectorSafely(currentElement, selector)))
+      matches.push(currentElement);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:827]: Adjudicated fail-closed: scope-start :scope requires an existing stylesheet scope context, which this resolver does not have. Added regression coverage for @scope (:scope > .theme) returning fallback.

File: packages/components/src/_internal/text-direction-css.ts [hunk 11]
```diff
@@ -163,6 +221,378 @@ function isConditionalRuleActive(
...
+      for (const index of path) {
+        const child = cloneElement.children[index];
+        if (!child) break;
+        cloneElement = child;
+      }
+      if (
+        Array.from(clone.querySelectorAll(selector.replace(/:scope\b/gi, `[${marker}]`))).includes(
+          cloneElement,
+        )
+      )
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:588]: Fixed in fbe88f53c: detached-clone matching now checks the marked clone root itself, with a regression forcing the clone fallback for :scope on the scope root.

File: packages/components/src/_internal/text-direction-css.ts [hunk 12]
```diff
@@ -48,6 +57,16 @@ function isActiveStyleSheet(sheet: CSSStyleSheet): boolean {
...
 }
 
+type ScopeRoot = Element | ShadowRoot;
+
+function getImplicitScopeRoot(sheet: CSSStyleSheet): ScopeRoot | null {
+  const ownerNode = Reflect.get(sheet, 'ownerNode');
+  if (!(ownerNode instanceof Element) || ownerNode.tagName !== 'STYLE') return null;
+  if (ownerNode.parentElement) return ownerNode.parentElement;
+  const root = ownerNode.getRootNode();
+  return root instanceof ShadowRoot ? root : null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in fbe88f53c by guarding every ShadowRoot reference with typeof ShadowRoot; typecheck and focused text-direction tests pass.

**Comment by stevekinney** [comment]
@copilot review please re-review exact head fbe88f53ce3bcfb75800928098ea56793a86d2d0, focusing on the resolved scope-boundary findings.

**Comment by stevekinney** [comment]
@codex review please re-review exact head fbe88f53ce3bcfb75800928098ea56793a86d2d0, focusing on the resolved scope-boundary findings.

File: packages/components/src/_internal/text-direction-css.ts [hunk 13]
```diff
@@ -163,6 +229,408 @@ function isConditionalRuleActive(
...
+      )
+        continue;
+      let matchedCloneElement: Element = cloneElement;
+      for (const index of path) {
+        const child = matchedCloneElement.children[index];
+        if (!child) break;
+        matchedCloneElement = child;
+      }
+      if (
+        cloneElement.matches(selector.replace(/:scope\b/gi, `[${marker}]`)) ||
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 0cc9bb9a0243e5fdef1cd965c54c008ee5849b51. The detached-clone path now accepts the clone root match only when the tested live element maps to that root; descendants must match through querySelectorAll. Regression: `does not match descendants for a cloned :scope selector` passes in the focused text-direction suite.

File: packages/components/src/_internal/text-direction-css.ts [hunk 14]
```diff
@@ -163,6 +229,408 @@ function isConditionalRuleActive(
...
+        Array.from(
+          cloneElement.querySelectorAll(selector.replace(/:scope\b/gi, `[${marker}]`)),
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 0cc9bb9a0243e5fdef1cd965c54c008ee5849b51. Replacement now tokenizes :scope while preserving quoted attribute text and escapes. Regression: `preserves quoted :scope text when replacing the pseudo-class` passes in the focused text-direction suite.

**Comment by stevekinney** [comment]
@copilot review\n\nPlease review exact head 0cc9bb9a0243e5fdef1cd965c54c008ee5849b51.

**Comment by stevekinney** [comment]
@codex review\n\nPlease review exact head 0cc9bb9a0243e5fdef1cd965c54c008ee5849b51.

File: packages/components/src/_internal/text-direction-css.ts [hunk 15]
```diff
@@ -163,6 +229,461 @@ function isConditionalRuleActive(
...
+}
+
+function findScopeMatches(element: HTMLElement, selectors: string[]): Element[] {
+  if (selectors.length === 0) return [];
+  const matches: Element[] = [];
+  let current: Element | null = element;
+  while (current) {
+    const currentElement: Element = current;
+    if (selectors.some((selector) => matchesSelectorSafely(currentElement, selector)))
+      matches.push(currentElement);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:827]: Fixed in 718408153. Exact `@scope (:scope)` roots now bind to the stylesheet implicit root and require the tested element to be contained by it; otherwise the scope fails closed. Regression `does not broaden an exact :scope scope-start beyond its stylesheet root` passes.

File: packages/components/src/_internal/text-direction-css.ts [hunk 16]
```diff
@@ -48,6 +57,24 @@ function isActiveStyleSheet(sheet: CSSStyleSheet): boolean {
   return matchMedia(mediaText).matches;
 }
 
+type ScopeRoot = Element | ShadowRoot;
+
+function getImplicitScopeRoot(sheet: CSSStyleSheet, fallbackRoot: Node): ScopeRoot | null {
+  const ownerNode = Reflect.get(sheet, 'ownerNode');
+  if (!(ownerNode instanceof Element) || ownerNode.tagName !== 'STYLE') {
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 718408153. Implicit scope owner detection uses case-insensitive `localName`, covering SVG style elements. Regression `recognizes SVG style elements as implicit scope owners` passes.

File: packages/components/src/_internal/text-direction-css.ts [hunk 17]
```diff
@@ -48,6 +57,24 @@ function isActiveStyleSheet(sheet: CSSStyleSheet): boolean {
...
+    return typeof ShadowRoot !== 'undefined' && fallbackRoot instanceof ShadowRoot
+      ? fallbackRoot
+      : null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 718408153. Stylesheet collection now carries each sheet’s actual adoption context, so document sheets do not inherit a target ShadowRoot while shadow-root sheets retain it. Existing adoption-root regression passes.

File: packages/components/src/_internal/text-direction-css.ts [hunk 18]
```diff
@@ -163,6 +229,461 @@ function isConditionalRuleActive(
...
+      while (current && current !== root) {
+        const parent: Element | null = current.parentElement;
+        if (!parent) {
+          if (
+            typeof ShadowRoot !== 'undefined' &&
+            root instanceof ShadowRoot &&
+            current.getRootNode() === root
+          )
+            break;
+          break;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:565]: Fixed in 718408153. ShadowRoot clone traversal records the direct child index before replaying the path through the wrapper, preserving target mapping. Shadow relative-selector regressions pass.

**Comment by stevekinney** [comment]
@copilot review

Please review exact head 718408153b0cc7fd8a2e83e7a82d95c25d924a20.

**Comment by stevekinney** [comment]
@codex review

Please review exact head 718408153b0cc7fd8a2e83e7a82d95c25d924a20.

**Comment by stevekinney** [comment]
@copilot review

Correction: review exact current head 7184081530ab9865d05adab469f1718b6ea95239.

**Comment by stevekinney** [comment]
@codex review

Correction: review exact current head 7184081530ab9865d05adab469f1718b6ea95239.

File: packages/components/src/_internal/text-direction-css.ts [hunk 19]
```diff
@@ -163,6 +232,474 @@ function isConditionalRuleActive(
...
+  prelude: ScopePrelude,
+  parentScopes: readonly ActiveScope[],
+  implicitScopeRoot: ScopeRoot | null,
+): ScopeRoot[] | null {
+  if (!selectorsAreValid(element, prelude.rootSelectors)) return null;
+  if (prelude.limitSelectors && !selectorsAreValid(element, prelude.limitSelectors)) return null;
+  if (prelude.rootSelectors.some(hasScopePseudoClass)) {
+    if (
+      !implicitScopeRoot ||
+      prelude.rootSelectors.some((selector) => selector.trim().toLowerCase() !== ':scope')
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 64e7d28b7 (initial fix in 8694a493e, corrected in 64e7d28b7).

`findActiveScopeRoots` now splits the prelude's root selectors into `scopeRootSelectors` (those with `:scope`) and `ordinaryRootSelectors` (plain selectors), and only bails with `null` when an unsupported scope-pseudo variant (anything other than exact `:scope`) has no ordinary selector to fall back on. For `@scope (:scope, .theme)`, `scopeRootSelectors = [':scope']` is exact, so it no longer rejects the whole rule — `.theme` is matched via `findScopeMatches` alongside the implicit scope root contributed by the exact `:scope` term.

Regression test: `retains ordinary roots in mixed :scope root lists` in `text-direction.test.ts`.

Note: while implementing this I found the first pass of the fix (8694a493e) let an *unsupported* scope-pseudo selector (e.g. `:scope > .foo`) contribute the implicit root whenever paired with any ordinary selector, even though the compound form itself was never evaluated — a false-positive risk. That's corrected in 64e7d28b7, which gates implicit-root inclusion on `scopeRootSelectors` being exactly `:scope`, with a new regression test (`does not activate the implicit root for an unsupported scope-pseudo selector`) that fails against 8694a493e and passes against 64e7d28b7.

File: packages/components/src/_internal/text-direction-css.ts [hunk 20]
```diff
@@ -10,29 +10,38 @@ export function matchesDirectionStyleRule(
...
   getParentElement: ParentElementResolver,
 ): boolean {
-  const styleSheets = new Set<CSSStyleSheet>([
-    ...Array.from(element.ownerDocument.styleSheets),
-    ...Array.from(element.ownerDocument.adoptedStyleSheets ?? []),
-  ]);
+  const styleSheets = new Map<CSSStyleSheet, ScopeRoot | null>();
+  for (const sheet of Array.from(element.ownerDocument.styleSheets)) styleSheets.set(sheet, null);
+  for (const sheet of Array.from(element.ownerDocument.adoptedStyleSheets ?? []))
+    styleSheets.set(sheet, null);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:16]: Fixed in 8694a493e.

`getImplicitScopeRoot` now takes `element.ownerDocument.documentElement` as a `documentRoot` parameter and returns it instead of `null` when the stylesheet has no owning `<style>` element and the fallback root isn't a `ShadowRoot` — i.e. document `<link>` sheets and constructed sheets in `document.adoptedStyleSheets`. Previously that path returned `null`, so an exact `:scope` root in one of those sheets could never resolve to a root and the rule was always treated as inactive.

Regression test: `uses the document root for exact :scope scopes in document stylesheets` in `text-direction.test.ts`, which mocks `document.styleSheets` with a sheet object that has no `ownerNode` (matching the existing test helper convention where other cases explicitly pass `ownerNode: styleElement`).

File: packages/components/src/_internal/text-direction-css.ts [hunk 21]
```diff
@@ -163,6 +232,474 @@ function isConditionalRuleActive(
...
+      let current: Element | null = element;
+      while (current && current !== root) {
+        const parent: Element | null = current.parentElement;
+        if (!parent) {
+          if (
+            typeof ShadowRoot !== 'undefined' &&
+            root instanceof ShadowRoot &&
+            current.getRootNode() === root
+          )
+            path.unshift(Array.prototype.indexOf.call(root.children, current));
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:564]: Fixed in 8694a493e.

`matchesScopedSelector` now computes the replaced selector once and, when the root is a `ShadowRoot` (represented for matching purposes by a synthetic `<div>` wrapper cloning its children, since `ShadowRoot` isn't an `Element` and `cloneNode` can't be used on it directly), checks whether the marker attribute is immediately followed by another simple selector with no combinator (`/^\[[^\]]+\][^\s>+~]/`) — i.e. a compound qualifier attached directly to `:scope`, such as `:scope:is(div)`. In that case it skips the root rather than testing the qualifier against the wrapper, because the wrapper is an arbitrary plain `<div>` with none of the real `ShadowRoot`'s identity — testing `:is(div)` against it would incorrectly pass (the wrapper literally is a div) regardless of what the actual shadow host is.

Regression test: `does not let a shadow clone wrapper satisfy root qualifiers` in `text-direction.test.ts`, which asserts `:scope:is(div) > .shell` inside a shadow root does NOT match (falls back to the inherited `rtl`).

File: packages/components/src/_internal/text-direction-css.ts [hunk 22]
```diff
@@ -163,6 +233,486 @@ function isConditionalRuleActive(
...
+    return false;
+  }
+}
+
+function matchesScopedSelector(
+  element: Element,
+  selector: string,
+  scopes: readonly ActiveScope[],
+): boolean {
+  if (!hasScopePseudoClass(selector)) return matchesSelectorSafely(element, selector);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 45419d9d3. Added `normalizeScopeRelativeSelector` in `matchesScopedSelector`: within an active scope, a selector containing a bare `&` that wasn't already resolved by `resolveNestedSelector` against a wrapping style rule (i.e. it reached here unchanged, meaning it sits directly under `@scope`/a conditional group rule with no intervening style rule) is rewritten to `:scope` before matching, reusing the existing scope-aware path instead of an ordinary `element.matches('&')` call.

Regression test: `treats a bare & as an alias for the scope root` in `text-direction.test.ts`. I verified it's discriminating by temporarily reverting the normalization and confirming the test fails (`Expected: "ltr", Received: "rtl"`) before restoring the fix.

[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 45419d9d3, same `normalizeScopeRelativeSelector` helper as the `&` fix (both are handled by the same normalization pass in `matchesScopedSelector`, since `matches()` rejects both). A selector that leads with a combinator (`>`, `+`, `~`) after any `&` substitution is prefixed with `:scope ` before matching.

Regression test: `treats a leading combinator as relative to the scope root`. Note this test uses a grandchild assertion specifically because this suite's DOM shim (happy-dom) doesn't actually throw on `element.matches('> .shell')` — it parses it leniently and matches `.shell` at any depth, which would make a naive same-position test pass for the wrong reason even without the fix. I confirmed the grandchild assertion is genuinely discriminating (fails with `Expected: "rtl", Received: "ltr"` when the fix is reverted).

File: packages/components/src/_internal/text-direction-css.ts [hunk 23]
```diff
@@ -163,6 +233,486 @@ function isConditionalRuleActive(
...
+  while (current) {
+    if (
+      limitSelectors.some((selector) =>
+        hasScopePseudoClass(selector)
+          ? matchesScopedSelector(current!, selector, [{ roots: [root] }])
+          : matchesSelectorSafely(current!, selector),
+      )
+    )
+      return true;
+    if (current === root) break;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 45419d9d3 in `isWithinScopeLimit`: the `current === root` break now runs before the limit-selector test, so the scoping root is never itself checked against `limitSelectors`. I generalized this beyond the `:scope`-referencing-limit repro in the review comment — the root is exempt from any limit selector, not just a `:scope` one, per the CSS Scoping spec ("no element matching L is both a *proper* descendant of R and an inclusive ancestor of E" — the root is never a proper descendant of itself). Concretely: `@scope (.card) to (.card) {}` — the root also matching its own limit — previously excluded the root's entire subtree; now it correctly stays in scope.

This also meant fixing a pre-existing test that encoded the old (incorrect) behavior: `respects @scope boundaries when scanning direction rules: outside the scope limit` asserted `'rtl'` for a root element carrying both the root and limit classes simultaneously. I renamed the case to `root also matching the scope limit stays in scope` and updated the expectation to `'ltr'`, with a comment explaining why.

Regression tests: the renamed case above, plus `does not create a limit element for a limit that references :scope` (the literal `@scope (.theme) to (:scope) {}` repro from the review). Both verified to fail without the fix.

File: packages/components/src/_internal/text-direction-css.ts [hunk 24]
```diff
@@ -163,6 +233,486 @@ function isConditionalRuleActive(
...
+          ...findScopeMatches(element, ordinaryRootSelectors),
+        ];
+  if (usesImplicitScopeRoot && implicitScopeRoot && !implicitScopeRoot.contains(element)) {
+    const implicitRootIndex = roots.indexOf(implicitScopeRoot);
+    if (implicitRootIndex >= 0) roots.splice(implicitRootIndex, 1);
+  }
+  if (prelude.rootSelectors.length === 0 && roots[0] && !roots[0].contains(element)) return null;
+  const activeRoots: ScopeRoot[] = [];
+  for (const root of roots) {
+    if (root && !isWithinParentScopes(root, parentScopes)) continue;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts]: Fixed in 45419d9d3. Added `scopePseudoRootCandidates(parentScopes, implicitScopeRoot)`: when `parentScopes.length > 0` it returns the nearest enclosing scope's active roots; only at top level does it fall back to the stylesheet's implicit root. An unrooted `@scope` now uses these candidates (filtered by `.contains(element)`) instead of unconditionally defaulting to `implicitScopeRoot ?? documentElement`, so `@scope (.outer) { @scope { .shell {...} } }` correctly intersects with `.outer` rather than falling back to the document root and failing the (unrelated) containment/parent-scope check.

Regression test: `binds an unrooted nested scope to the enclosing scope root, not the document`, verified to fail (`Expected: "ltr", Received: "rtl"`) without the fix. I also re-verified the two existing tests most likely to interact with this change still pass for the right reason (`fails closed when an implicit scope is excluded by an enclosing scope` — outer scope never activates, so the inner unrooted scope is never even reached; `requires nested scopes to intersect every enclosing scope` — a rooted, not unrooted, inner scope, so this code path doesn't apply).

File: packages/components/src/_internal/text-direction-css.ts [hunk 25]
```diff
@@ -163,6 +233,486 @@ function isConditionalRuleActive(
...
+  if (!selectorsAreValid(element, prelude.rootSelectors)) return null;
+  if (prelude.limitSelectors && !selectorsAreValid(element, prelude.limitSelectors)) return null;
+  const scopeRootSelectors = prelude.rootSelectors.filter(hasScopePseudoClass);
+  const ordinaryRootSelectors = prelude.rootSelectors.filter(
+    (selector) => !hasScopePseudoClass(selector),
+  );
+  const unsupportedScopeRoot = scopeRootSelectors.some(
+    (selector) => selector.trim().toLowerCase() !== ':scope',
+  );
+  if (unsupportedScopeRoot && ordinaryRootSelectors.length === 0) return null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:408]: Partially fixed in 45419d9d3; the rest is filed as #1138 with a verbatim repro.

**Fixed here:** the exact-`:scope` half. `scopePseudoRootCandidates` (same helper as the (d) fix above) now resolves an exact `:scope` scope-start selector against the nearest enclosing scope's roots when nested, instead of always using the stylesheet's implicit root — so `@scope (.parent) { @scope (:scope) { .shell {...} } }` correctly binds to `.parent`. Regression test: `binds a nested exact :scope root to the enclosing scope, not the stylesheet root`, verified to fail without the fix.

**Deferred to #1138:** the relative-selector half (`@scope (:scope > .child) {...}` when nested) still fails closed via the existing `unsupportedScopeRoot` guard — I did not remove that guard, since it's also the fix for two other already-resolved threads on this PR (rejecting unsupported scope-pseudo roots, and the mixed-list case). Making `:scope > .child` work requires generalizing root *resolution* to walk candidate ancestors and test each one with `matchesScopedSelector` against the parent scope, analogous to what `findScopeMatches` does for ordinary selectors — a real algorithm change I didn't want to bundle into this already-large merge+fix push. Repro is in #1138.

File: packages/components/src/_internal/text-direction-css.ts [hunk 26]
```diff
@@ -163,6 +233,486 @@ function isConditionalRuleActive(
...
+  for (const root of scope.roots) {
+    try {
+      if (root.querySelector(':scope') === root) {
+        if (Array.from(root.querySelectorAll(selector)).includes(element)) return true;
+        continue;
+      }
+      const clone =
+        typeof ShadowRoot !== 'undefined' && root instanceof ShadowRoot
+          ? null
+          : root.cloneNode(true);
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:533]: Confirmed real; deferred to #1138 (filed with a verbatim repro) rather than fixed in this push.

The detached-clone fallback in `matchesScopedSelector` only clones the scope root's own subtree, so a selector relating the root to an ancestor outside the root (e.g. `main :scope .shell` for `<main><section class="theme"><div class="shell"></div></section></main>`) can never match — the clone has no `main` to test against.

I looked at fixing this directly but it isn't a contained change: the mutation-free clone strategy here exists specifically because an earlier resolved thread on this PR required *not* mutating the live scope root (observed elements were getting spurious mutation records). Preserving outside-ancestor context means either reconstructing enough of the real ancestor chain in the clone (tag names, classes, sibling position for things like `:nth-child`) or a different matching strategy entirely — a real design decision, not a same-PR fix. Full repro and reasoning are in #1138.

File: packages/components/src/_internal/text-direction-css.ts [hunk 27]
```diff
@@ -262,6 +332,545 @@ function isConditionalRuleActive(
...
+    );
+  }
+  const scopeRootSelectors = prelude.rootSelectors.filter(hasScopePseudoClass);
+  const ordinaryRootSelectors = prelude.rootSelectors.filter(
+    (selector) => !hasScopePseudoClass(selector),
+  );
+  const unsupportedScopeRoot = scopeRootSelectors.some(
+    (selector) => selector.trim().toLowerCase() !== ':scope',
+  );
+  if (unsupportedScopeRoot && ordinaryRootSelectors.length === 0) return null;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:408]: Real, and fail-closed (the direction hint goes unapplied; no incorrect direction is ever applied — the module's stated design premise). Absorbed into the @scope spec-parity backlog #1138 as a refinement of the mixed-list/relative-selector work tracked there — see https://github.com/stevekinney/cinder/issues/1138 for the full item list. This PR's drain round fixed the 4 tractable @scope findings with regression tests (45419d9d3); these two sit in the same code path as the deferred relative-form root-candidate rework and should be fixed together with it.

File: packages/components/src/_internal/text-direction-css.ts [hunk 28]
```diff
@@ -262,6 +332,545 @@ function isConditionalRuleActive(
...
+// Rewrites the two relative-selector shorthands `@scope` defines into an
+// explicit `:scope`, so both reuse the scope-aware matching in
+// `matchesScopedSelector` instead of an ordinary (and, for the
+// leading-combinator form, invalid) `element.matches()` call:
+//  - `&`, when it wasn't already resolved by `resolveNestedSelector` against
+//    a wrapping style rule, is CSS's own alias for the scope root.
+//  - A selector that leads with a combinator (`> .foo`, `+ .foo`, `~ .foo`)
+//    is shorthand for that combinator applied to the scope root.
+function normalizeScopeRelativeSelector(selector: string): string {
+  if (!selector.includes('&') && !/^\s*[>+~]/.test(selector)) return selector;
```
[the remark this answers is not in the document]
Comment (stevekinney) [reply packages/components/src/_internal/text-direction-css.ts:496]: Real, and fail-closed (the direction hint goes unapplied; no incorrect direction is ever applied — the module's stated design premise). Absorbed into the @scope spec-parity backlog #1138 as a refinement of the mixed-list/relative-selector work tracked there — see https://github.com/stevekinney/cinder/issues/1138 for the full item list. This PR's drain round fixed the 4 tractable @scope findings with regression tests (45419d9d3); these two sit in the same code path as the deferred relative-form root-candidate rework and should be fixed together with it.
